src/plugins/workspace/public/plugin.ts

```diff
@@ -2,78 +2,79 @@
+import { NavLinkWrapper } from '../../../core/public/chrome/nav_links/nav_link';
 
 type WorkspaceAppType = (params: AppMountParameters, services: Services) => () => void;
 
 export class WorkspacePlugin implements Plugin<{}, {}> {
   private coreStart?: CoreStart;
+  private currentWorkspaceIdSubscription?: Subscription;
   private currentWorkspaceSubscription?: Subscription;
   private getWorkspaceIdFromURL(): string | null {
     return getWorkspaceIdFromUrl(window.location.href);
   }
 
-  private filterByWorkspace(workspace: WorkspaceObject | null, allNavLinks: ChromeNavLink[]) {
-    if (!workspace) return allNavLinks;
-    const features = workspace.features ?? ['*'];
-    return allNavLinks.filter(featureMatchesConfig(features));
-  }
+  /**
+   * Filter the nav links based on the feature configuration of workspace
+   */
+  private filterByWorkspace(allNavLinks: NavLinkWrapper[], workspace: WorkspaceObject | null) {
+    if (!workspace || !workspace.features) return allNavLinks;
 
-  private filterNavLinks(core: CoreStart) {
-    const navLinksService = core.chrome.navLinks;
-    const allNavLinks$ = navLinksService.getAllNavLinks$();
-    const currentWorkspace$ = core.workspaces.currentWorkspace$;
-    combineLatest([
-      allNavLinks$.pipe(map(this.changeCategoryNameByWorkspaceFeatureFlag)),
-      currentWorkspace$,
-    ]).subscribe(([allNavLinks, currentWorkspace]) => {
-      const filteredNavLinks = this.filterByWorkspace(currentWorkspace, allNavLinks);
-      const navLinks = new Map<string, ChromeNavLink>();
-      filteredNavLinks.forEach((chromeNavLink) => {
-        navLinks.set(chromeNavLink.id, chromeNavLink);
-      });
-      navLinksService.setNavLinks(navLinks);
-    });
+    const featureFilter = featureMatchesConfig(workspace.features);
+    return allNavLinks.filter((linkWrapper) => featureFilter(linkWrapper.properties));
   }
 
   /**
-   * The category "Opensearch Dashboards" needs to be renamed as "Library"
-   * when workspace feature flag is on, we need to do it here and generate
-   * a new item without polluting the original ChromeNavLink.
+   * Filter nav links by the current workspace, once the current workspace change, the nav links(left nav bar)
+   * should also be updated according to the configured features of the current workspace
    */
-  private changeCategoryNameByWorkspaceFeatureFlag(chromeLinks: ChromeNavLink[]): ChromeNavLink[] {
-    return chromeLinks.map((item) => {
-      if (item.category?.id === DEFAULT_APP_CATEGORIES.opensearchDashboards.id) {
-        return {
-          ...item,
-          category: {
-            ...item.category,
-            label: i18n.translate('core.ui.libraryNavList.label', {
-              defaultMessage: 'Library',
-            }),
-          },
-        };
-      }
-      return item;
+  private filterNavLinks(core: CoreStart) {
+    const currentWorkspace$ = core.workspaces.currentWorkspace$;
+    let filterLinksByWorkspace: LinksUpdater;
```

Comment:
will this `filterLinksByWorkspace` always be undefined, if switch to another workspace and whole page get refreshed?